logback/src/main/java/com/linecorp/armeria/common/logback/UnionMap.java

```diff
@@ -108,6 +110,99 @@ public void clear() {
             return entrySet;
         }
 
-        return entrySet = Collections.unmodifiableSet(Sets.union(first.entrySet(), second.entrySet()));
+        // Check for dupes first to reduce allocations on the vastly more common case where there aren't any.
+        boolean secondHasDupes = false;
+        final Set<Entry<K, V>> secondEntries = second.entrySet();
+        for (Entry<K, V> entry : secondEntries) {
+            if (first.containsKey(entry.getKey())) {
+                secondHasDupes = true;
+                break;
+            }
+        }
+
+        final Set<Entry<K, V>> filteredSecond;
+        if (!secondHasDupes) {
+            filteredSecond = secondEntries;
+        } else {
+            filteredSecond = new LinkedHashSet<>();
+            for (Entry<K, V> entry : secondEntries) {
+                if (!first.containsKey(entry.getKey())) {
+                    filteredSecond.add(entry);
+                }
+            }
+        }
+        return entrySet = Collections.unmodifiableSet(new ConcatenatedSet<>(first.entrySet(), filteredSecond));
```

Comment:
This is the last comment I believe: Do we need to wrap `ConcatenatedSet` with `UnmodifiableSet`? It looks unmodifiable already.